### PR TITLE
Selenium testing

### PR DESCRIPTION
**Pull request by delterr**
This PR aims to add in testing to the current tech-audit-tool UI. The main goal of this is to automate the testing process to save time during development.

The main test is the project creation test as that takes up the most time during development which can be found in `tests/test_project_creation.py`. The other tests are for testing signing in and basic page navigation.

Other changes: Minor changes to page rendering, fixed session expiry bug.

This PR uses `selenium ^4.25.0`. Make sure to first  `poetry install` before running any tests.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -242,4 +242,28 @@ Use the make command in the root directory of the project to run the app:
...
 ```
 
-This will run `isort`, `black` and `flake8`. Flake8 will ignore `E501 line too long`.
\ No newline at end of file
+This will run `isort`, `black` and `flake8`. Flake8 will ignore `E501 line too long`.
+
+## Testing
+### Setting up test environment variables
+
+```bash
```
Comment (TotalDwarf03) [inline README.md:250]: Worth mentioning what these env variables are for and providing a sensible example

File: tests/test_login.py
Comment (TotalDwarf03) [inline tests/test_login.py:1]: Can we add docstrings to this please? There are some good vscode extension to make this quicker. I use [this](https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring). 

I'd also recommend following [Google's format](https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings) of docstring.

File: tests/test_navigation.py
Comment (TotalDwarf03) [inline tests/test_navigation.py:1]: Docstrings

File: tests/test_project_creation.py
Comment (TotalDwarf03) [inline tests/test_project_creation.py:1]: Docstrings

File: README.md [hunk 2]
```diff
@@ -242,4 +242,28 @@ Use the make command in the root directory of the project to run the app:
...
+export TEST_EMAIL=<email>
+export TEST_PASSWORD=<password>
+```
+
+## Running tests
```
Comment (TotalDwarf03) [inline README.md]: nit: Capitalise Titles

**Review by TotalDwarf03 — COMMENTED** [review]
Couple of small things. Going to try and run the project locally.

**Comment by TotalDwarf03** [comment]
All working locally. Nice to actually use the tool for the first time. I have spotted one or two minors but will discuss in standup Wednesday. Happy to approve once the comments above are sorted.

File: README.md (hunk 1 above)
Comment (delterr) [reply README.md:250]: > Worth mentioning what these env variables are for and providing a sensible example

Fixed

File: tests/test_login.py
Comment (delterr) [reply tests/test_login.py:1]: > Can we add docstrings to this please? There are some good vscode extension to make this quicker. I use [this](https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring).
> 
> I'd also recommend following [Google's format](https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings) of docstring.

Fixed

File: tests/test_navigation.py
Comment (delterr) [reply tests/test_navigation.py:1]: > Docstrings

Fixed

File: tests/test_project_creation.py
Comment (delterr) [reply tests/test_project_creation.py:1]: > Docstrings

Fixed

File: README.md (hunk 2 above)
Comment (delterr) [reply README.md]: > nit: Capitalise Titles

Fixed

File: Makefile [hunk 3]
```diff
@@ -30,4 +33,7 @@ docker-run:
 	docker run -p 127.0.0.1:8000:8000 -e AWS_SECRET_ACCESS_KEY -e AWS_ACCESS_KEY_ID \
 	 -e API_BUCKET_NAME -e API_SECRET_NAME \
 	  -e UI_SECRET_NAME -e AWS_ACCOUNT_NAME \
-	  -e LOCALHOST tech-audit-tool
\ No newline at end of file
+	  -e LOCALHOST tech-audit-tool
+
+test:
+	cd tests && poetry run python3 test_login.py
```
Comment (sebtheo) [inline Makefile]: Could you add other make commands that ran all the tests?

File: README.md [hunk 4]
```diff
@@ -242,4 +242,27 @@ Use the make command in the root directory of the project to run the app:
...
+The following test will automatically go through the steps required for creating a project.
+This is useful if you wish to quickly populate the fields after a change has been made instead of populating the fields manually.
+```bash
+make test-project-creation
+```
+
+### Login Test
+```bash
+poetry run python3 tests/test_login.py
+```
```
Comment (sebtheo) [inline README.md:267]: Maybe add more information about what tests there are as there's only a few.

File: application.py
Comment (sebtheo) [inline application.py:1]: You could create a helper function that `get_id_token()` then place the try and except in there so you don't have to repeat the try and except so much?

File: tests/test_login.py
Comment (sebtheo) [inline tests/test_login.py:1]: Add logging to the tests? When I run the test it goes super fast and when it sometimes errors I don't know at what point it errored.

File: tests/test_project_creation.py
Comment (sebtheo) [inline tests/test_project_creation.py:1]: I got an error when I ran this. It went all the way to the end, then it took me back to dashboard, then it said an error. Maybe we can jump on a call so I can show you?

Also, it enters in the same data each time. Surely it won't work as the same project name is being applied each time? There are libraries that you can use that generate fake information such as first names, emails and probably project names (could just do 3 random words).

File: tests/test_utils.py
Comment (sebtheo) [inline tests/test_utils.py:1]: This doesn't call the process at the bottom so nothing happens.

```
if __name__ == "__main__":
    unittest.main()
```

**Review by sebtheo — DISMISSED** [review]
I would add logging for all tests. Then a user can see what is going on as they go through it and can identify at what points there are errors.

Random/fake information when creating a project?

I added the part in the login test to save cookies so that on each test you didn't have to login, you could load on to dashboard and start the test. Maybe the login is a one time thing.

Also you have the ability to make a selenium browser headless so it still interacts with the elements on the page still but it doesn't show the browser on the user's computer, like this:

```
options = Options()
options.add_argument('--headless=new')
driver = webdriver.Chrome(path, options=options)
```

**Comment by delterr** [comment]
> I would add logging for all tests. Then a user can see what is going on as they go through it and can identify at what points there are errors.
> 
> Random/fake information when creating a project?
> 
> I added the part in the login test to save cookies so that on each test you didn't have to login, you could load on to dashboard and start the test. Maybe the login is a one time thing.
> 
> Also you have the ability to make a selenium browser headless so it still interacts with the elements on the page still but it doesn't show the browser on the user's computer, like this:
> 
> ```
> options = Options()
> options.add_argument('--headless=new')
> driver = webdriver.Chrome(path, options=options)
> ```

I'll be honest, I don't think the test_login is even necessary since we're testing all that during project creation. Same goes with the test_navigation test, its all basically done in the project creation tests

**Comment by sebtheo** [comment]
> > I would add logging for all tests. Then a user can see what is going on as they go through it and can identify at what points there are errors.
> > Random/fake information when creating a project?
> > I added the part in the login test to save cookies so that on each test you didn't have to login, you could load on to dashboard and start the test. Maybe the login is a one time thing.
> > Also you have the ability to make a selenium browser headless so it still interacts with the elements on the page still but it doesn't show the browser on the user's computer, like this:
> > ```
> > options = Options()
> > options.add_argument('--headless=new')
> > driver = webdriver.Chrome(path, options=options)
> > ```
> 
> I'll be honest, I don't think the test_login is even necessary since we're doing testing all that during project creation. Same goes with the test_navigation test, its all basically done in the project creation tests

Yeah my idea was having a user/tester logging in once to setup their cookies. Then when they run through the other tests then it doesn't log in, it just uses the cookies that have been saved.

Either remove test_login.py or do the cookies things, I guess?

Each test they make they are getting a new id_token and it's an extra process/step. Although, if they test, then come back after the token has expired then they have to test_login again to get new cookies, instead of logging in fresh each time.

Either way is fine of course, I'm just being annoying I think. @delterr

File: tests/test_project_creation.py
Comment (delterr) [reply tests/test_project_creation.py:1]: Added a bunch of implicit waits before find_element commands, hopefully that fixes the issue

File: tests/test_utils.py
Comment (delterr) [reply tests/test_utils.py:1]: Could you clarify this comment a little more?

**Comment by delterr** [comment]
> > > I would add logging for all tests. Then a user can see what is going on as they go through it and can identify at what points there are errors.
> > > Random/fake information when creating a project?
> > > I added the part in the login test to save cookies so that on each test you didn't have to login, you could load on to dashboard and start the test. Maybe the login is a one time thing.
> > > Also you have the ability to make a selenium browser headless so it still interacts with the elements on the page still but it doesn't show the browser on the user's computer, like this:
> > > ```
> > > options = Options()
> > > options.add_argument('--headless=new')
> > > driver = webdriver.Chrome(path, options=options)
> > > ```
> > 
> > 
> > I'll be honest, I don't think the test_login is even necessary since we're doing testing all that during project creation. Same goes with the test_navigation test, its all basically done in the project creation tests
> 
> Yeah my idea was having a user/tester logging in once to setup their cookies. Then when they run through the other tests then it doesn't log in, it just uses the cookies that have been saved.
> 
> Either remove test_login.py or do the cookies things, I guess?
> 
> Each test they make they are getting a new id_token and it's an extra process/step. Although, if they test, then come back after the token has expired then they have to test_login again to get new cookies, instead of logging in fresh each time.
> 
> Either way is fine of course, I'm just being annoying I think. @delterr

When I was creating the tests I wanted it to be an "end-to-end" test where the user logs in, navigates through to project creation and creates the project; So I don't believe its necessary to load the cookies in as logging in is part of the test (reason I called it "end to end")

File: Makefile
Comment (gibbardsteve) [inline Makefile:1]: make test: refers to make test-navigation, but this doesn't exist

File: tests/test_navigation.py
Comment (gibbardsteve) [inline tests/test_navigation.py:1]: Doesn't seem to be run in the makefile

File: tests/test_project_creation.py [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+        Args:
+            num_words (int): Number of words to be generated
+            delimiter (str, optional): Defines how to seprate words. Defaults to " ".
+
+        Returns:
+            str: Final string of words
+        """
+        return delimiter.join(list(map(str.title, [requests.get("https://random-word-api.herokuapp.com/word").json()[0] for word in range(num_words)])))
+
+    def test_project_details(self):
```
Comment (gibbardsteve) [inline tests/test_project_creation.py:55]: Can we also test the successful retrieval of the created project from the API by checking it exists in View your Projects and that the data in the summary matches that which was entered.

**Review by gibbardsteve — COMMENTED** [review]
Can we also increase the app log level from DEBUG to Warning, alternatively ensure that botocore/boto3 is not logging at DEBUG level

**Comment by gibbardsteve** [comment]
One other thing (we can add this as a separate ticket - not for this review). I'd like testing like this to use Behaviour Driven Development where the tests have user digestible description above the selenium. Look at gherkin/cucmber and feature files.  This allows you to specify test cases that are easily readable:

Feature: Project Creation 
  Scenario: Successful summary of project details
    Given the user has entered valid project details  
    When the user navigates to the summary page 
    Then the summary page should display the entered project details
 
 Writing tests this way mirrors the user requirements and makes it easier to see what testing we have and where ther are gaps.  Can you open a JIRA ticket in the back log please?

File: Makefile
Comment (delterr) [reply Makefile:1]: > make test: refers to make test-navigation, but this doesn't exist

Fixed

File: tests/test_navigation.py
Comment (delterr) [reply tests/test_navigation.py:1]: > Doesn't seem to be run in the makefile

Fixed

File: tests/test_project_creation.py (hunk 5 above)
Comment (delterr) [reply tests/test_project_creation.py:55]: > Can we also test the successful retrieval of the created project from the API by checking it exists in View your Projects and that the data in the summary matches that which was entered.

Added

File: tests/test_project_creation.py [hunk 6]
```diff
@@ -105,6 +105,10 @@ def test_project_details(self):
         button = driver.find_element(By.ID, 'submit-button')
         driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")
         self.wait.until(EC.element_to_be_clickable(button)).click()
+        
+        assert driver.current_url == "http://localhost:8000/dashboard"
+        assert self.project_name in driver.find_element(By.ID, "my-projects").text
```
Comment (gibbardsteve) [inline tests/test_project_creation.py:110]: Can we also validate the data displayed on the summary screen is as expected?  This is doing a fetch of the data from the API so is useful to validate that we don't have issues that side of the fence too.

**Review by gibbardsteve — COMMENTED** [review]
One comment that I don't think was fully resolved, other than that looks good.

File: tests/test_project_creation.py (hunk 6 above)
Comment (delterr) [reply tests/test_project_creation.py:110]: > Can we also validate the data displayed on the summary screen is as expected? This is doing a fetch of the data from the API so is useful to validate that we don't have issues that side of the fence too.

Fixed

**Review by gibbardsteve — APPROVED** [review]
Looks good - thanks for making the change

**Pull request merged**